File: greedy/2167. Minimum-Time-to-Remove-All-Cars-Containing-Illegal-Goods/2167. Minimum-Time-to-Remove-All-Cars-Containing-Illegal-Goods.py
class Solution:
    def minimumTime(self, s: str) -> int:
        left, n, res = 0, len(s), len(s)

        for i, char in enumerate(s):
            if char == '1':
                left = min(i + 1, left + 2)

            res = min(res, left + n - (i + 1))

        return res


x = Solution()
print(x.minimumTime('010101'))

File: greedy/2167. Minimum-Time-to-Remove-All-Cars-Containing-Illegal-Goods/2167. Minimum-Time-to-Remove-All-Cars-Containing-Illegal-Goods.ts
/**
 * Time complexity: O(n)
 * Space complexity: O(1)
 *
 * @param s
 */
function minimumTimeVersion2(s: string): number {
  const n = s.length
  let min = 1e10,
    left = 0

  for (let i = 0; i < n; i++) {
    if (s.charAt(i) === '1')
      left = Math.min(i + 1, left + 2)
    min = Math.min(min, left + n - (i + 1))
  }

  return min
}
